platform support documentation [skip ci] (#1605)
open-quantum-safe/liboqs: PLATFORMS.md

# Supported platforms

This file documents the different platforms supported by `liboqs` and therefore defines three different support tiers:

## Support tiers

> This classification is roughly based on the [rust platform support tier classification](https://doc.rust-lang.org/beta/rustc/platform-support.html):

Reviewer: Might I suggest either a "Tier 0" or additional criteria for "Tier 1" which includes constant-time testing and documentation of CT issues? We currently have automated CT-testing for x86 but not aarch64. Or would this be viewed as out of scope for this document?

Author: Sounds reasonable. Would you want to add proposed text straight to this PR?

Reviewer 2: @SWilson4 will you do this update?

Reviewer 2: Oh never mind I see that this was done, thanks.

The availability or tier of a target in stable `liboqs` is not a hard stability guarantee about the future availability or tier of that target. Higher-level target tiers are an increasing commitment to the support of a target, and we will take that commitment and potential disruptions into account when evaluating the potential demotion or removal of a target that has been part of a stable release. The promotion or demotion of a target will not generally affect existing stable releases, only current development and future releases.

In this policy, the words "must" and "must not" specify absolute requirements that a target must meet to qualify for a tier. The words "should" and "should not" specify requirements that apply in almost all cases, but for which the approving teams may grant an exception for good reason. The word "may" indicates something entirely optional, and does not indicate guidance or recommendations. This language is based on [IETF RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

## Platforms supported

### Tier 1

- x86_64/amd64/x64 for Ubuntu Linux (Focal) and MacOS (XCode 14)
- aarch64 for Ubuntu (Focal)

> - armhf/ARM7 and aarch64 emulation on Ubuntu

Reviewer 2: If armhf is at Tier 1 (or will be with the ci-containers update), great!  Otherwise okay to classify it as tier 2.

Author: We have [a successful CI run](https://github.com/open-quantum-safe/liboqs/actions/runs/6776860848/job/18419347065) as I already did the [ci-containers update (adding doxygen)](https://github.com/open-quantum-safe/ci-containers/pull/69) -- without PR review, admittedly: I didn't want to wait another 24 h to see the HQC CI pass.

> - ppc64le for Ubuntu (Focal)

Reviewer: I didn't recall that we had CI tests for ppc and s390x, so I dug around and found the Travis CI setup. But it looks like it hasn't been running in a while: https://app.travis-ci.com/github/open-quantum-safe/liboqs. Unless I'm missing something, we should probably knock these two down to Tier 3 pending a CI fix.

Reviewer: > _Very_ interesting. Didn't realize that (either). Makes perfect sense -- but I suppose you mean knocking them down to Tier 2?

I had thought Tier 3 since it doesn't look to me like we're even building on those architectures anywhere at the moment (until Travis is up and running again).

- s390x for Ubuntu (Focal)

### Tier 2

- x86_64/amd64/x64 for Windows 2022
- armeabi-v7a, arm64-v8a, x86, x86_64 for Android

> - OS64, TVOS for Apple Mobile

Reviewer 2: I don't know what OS64 is?

Author: Me neither. It's [this test run](https://github.com/open-quantum-safe/liboqs/actions/runs/6776860846/job/18419274742) triggered by https://github.com/open-quantum-safe/liboqs/blob/aeac3a441dd3d95821b04a26baae6f73013538d8/.github/workflows/apple.yml#L12 contributed in https://github.com/open-quantum-safe/liboqs/pull/1546. @res0nance could you clarify?.

Reviewer 3: OS64 is ARM64 for iphone

### Tier 3

- x86 for Windows

> - aarch64 for MacOS

Reviewer 2: aarch64 is clearly the intended future of the macOS platform, so we would definitely want to get this up to tier 1.  Is it not higher right now because we don't have any CI on this platform?

Author: >  Is it not higher right now because we don't have any CI on this platform

Correct

Reviewer 4: Github now has M1 runners I believe. I learnt this when someone was complaining about how expensive they are, though, and I'm not sure if they are available for OSS yet.

Author: > I'm not sure if they are available for OSS yet.

They are not: See [documentation](https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners#machine-specs-for-larger-runners) and [worked example](https://github.com/open-quantum-safe/oqs-provider/actions/runs/7075061285/job/19256776098).

Author: For completeness: Error message:

```
The job was not started because recent account payments have failed or your spending limit needs to be increased. Please check the 'Billing & plans' section in your settings.
```

Reviewer: I agree that aarch64/Mac is correctly classified as "tier 3" but ideally would be higher. Perhaps we could look into using our M1 profiling machine as a self-hosted runner? I don't think we would want the strain of running tests on it as often as on the GitHub runners, but a scheduled daily or weekly job would be more palatable.

Author: Agreed. Also thought the same. But wouldn't this effort be a waste of time? The LF companies will pay hundreds of thousands -- why not use (a tiny fraction) of this money to pay for M1 runners and save us this rather tedious work? It surely would be a better use of the money than for CO2-gobbling trips of some LF executives to PQ "workshops" :)

Reviewer 2: I've created issue #1630 to track this, and will close this thread as resolved.

# Supported platforms

This file documents the different platforms supported by `liboqs` and therefore defines three different support tiers:

## Support tiers

This classification is roughly based on the [rust platform support tier classification](https://doc.rust-lang.org/beta/rustc/platform-support.html):

### Tier 1

> Tier 1 targets can be thought of as "guaranteed to work". The CI system builds and tests binary versions for each tier 1 target to make sure any change does not negatively affect those platforms. Platform-specific build documentation must exist. Tier 1 targets marked with an asterisk are additionally tested for constant-time behaviour. The CI system contains automated constant-time testing for each of these starred targets, and all failures are documented in the `tests/constant_time` directory. IMPORTANT: This does not mean that constant-time behaviour is guaranteed on these targets, or that non-constant-time behaviour is limited to documented exceptions. It does, however, mean that `liboqs` developers should track constant-time issues on these platforms.

Reviewer: ```suggestion
Tier 1 targets can be thought of as "guaranteed to work". The CI system builds and tests binary versions for each tier 1 target to make sure any change does not negatively affect those platforms. Platform-specific build documentation must exist. Tier 1 targets marked with a dagger (†) are additionally tested for constant-time behaviour. The CI system contains automated constant-time testing for each of these starred targets, and all failures are documented in the `tests/constant_time` directory. IMPORTANT: This does not mean that constant-time behaviour is guaranteed on these targets, or that non-constant-time behaviour is limited to documented exceptions. It does, however, mean that `liboqs` developers should track constant-time issues on these platforms.
```

The availability or tier of a target in stable `liboqs` is not a hard stability guarantee about the future availability or tier of that target. Higher-level target tiers are an increasing commitment to the support of a target, and we will take that commitment and potential disruptions into account when evaluating the potential demotion or removal of a target that has been part of a stable release. The promotion or demotion of a target will not generally affect existing stable releases, only current development and future releases.

In this policy, the words "must" and "must not" specify absolute requirements that a target must meet to qualify for a tier. The words "should" and "should not" specify requirements that apply in almost all cases, but for which the approving teams may grant an exception for good reason. The word "may" indicates something entirely optional, and does not indicate guidance or recommendations. This language is based on [IETF RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

## Platforms supported

### Tier 1

> - x86_64/amd64/x64 for Ubuntu Linux (Focal)*

Reviewer: ```suggestion
- x86_64/amd64/x64 for Ubuntu Linux (Focal)†
```
